src/met.hxx

```diff
@@ -0,0 +1,389 @@
+    const std::string &genparticle_eta, const std::string &genparticle_phi,
+    const std::string &genparticle_mass, const std::string &genparticle_id,
+    const std::string &genparticle_status, const std::string &jet_pt,
+    const std::string &outputname, const std::string &recoilfile,
+    const std::string &systematicsfile, bool applyRecoilCorrections,
+    bool resolution, bool response, bool shiftUp, bool shiftDown) {
+    if (applyRecoilCorrections) {
+        Logger::get("RecoilCorrections")->debug("Will run recoil corrections");
+        const auto corrector = new RecoilCorrector(recoilfile);
+        const auto systematics = new METSystematic(systematicsfile);
+        auto shiftType = METSystematic::SysShift::Nominal;
+        if (shiftUp) {
+            shiftType = METSystematic::SysShift::Up;
+        } else if (shiftDown) {
+            shiftType = METSystematic::SysShift::Down;
+        }
+        auto sysType = METSystematic::SysType::None;
+        if (response) {
+            sysType = METSystematic::SysType::Response;
+        } else if (resolution) {
+            sysType = METSystematic::SysType::Resolution;
+        }
+        auto RecoilCorrections = [sysType, systematics, shiftType, corrector](
+                                     ROOT::Math::PtEtaPhiMVector &met,
+                                     const ROOT::RVec<float> &genparticle_pt,
+                                     const ROOT::RVec<float> &genparticle_eta,
+                                     const ROOT::RVec<float> &genparticle_phi,
+                                     const ROOT::RVec<float> &genparticle_mass,
+                                     const ROOT::RVec<int> &genparticle_id,
+                                     const ROOT::RVec<int> &genparticle_status,
+                                     const ROOT::RVec<float> &jet_pt) {
+            // TODO is this the correct number of jets ?
+            auto jets_above30 =
+                Filter(jet_pt, [](float pt) { return pt > 30; });
+            int nJets30 = jets_above30.size();
+            float MetX = met.Px();
+            float MetY = met.Py();
+            float correctedMetX = 0.;
+            float correctedMetY = 0.;
+            ROOT::Math::PtEtaPhiMVector genparticle;
+            ROOT::Math::PtEtaPhiMVector corrected_met;
+            float genPx = 0.; // generator Z(W) px
+            float genPy = 0.; // generator Z(W) py
+            float visPx = 0.; // visible (generator) Z(W) px
+            float visPy = 0.; // visible (generator) Z(W) py
+            // now loop though all genparticles in the event
+            for (std::size_t index = 0; index < genparticle_id.size();
+                 ++index) {
+                // consider a genparticle,
+                // 1. if it is a lepton and fromHardProcess --> bit 8 from
+                // status
+                // 2. if it is isDirectHardProcessTauDecayProduct --> bit 10
+                // in status
```

Comment:
We will need the same procedure of building the gen V for the Z pt mass reweighting. We should think about putting this into an extra function, maybe even an extra producer and df column.

Reply:
hmm shot, i even had it as an extra function before

Reply:
done